File: 3772-minimum-pair-removal-to-sort-array-ii/minimum-pair-removal-to-sort-array-ii.java
class Solution {
    class Pair implements Comparable<Pair> {
        long sum;
        int index;

        Pair(long sum, int index) {
            this.sum = sum;
            this.index = index;
        }

        public int compareTo(Pair other) {
            if (this.sum != other.sum) {
                return Long.compare(this.sum, other.sum);
            }
            return Integer.compare(this.index, other.index);
        }
    }

    public int minimumPairRemoval(int[] nums) {
        int n = nums.length;
        if (n <= 1) return 0;

        long[] val = new long[n];
        int[] prev = new int[n];
        int[] next = new int[n];
        boolean[] removed = new boolean[n];

        for (int i = 0; i < n; i++) {
            val[i] = nums[i];
            prev[i] = i - 1;
            next[i] = i + 1;
        }
        next[n - 1] = -1;

        PriorityQueue<Pair> pq = new PriorityQueue<>();
        int inversions = 0;

        for (int i = 0; i < n - 1; i++) {
            pq.add(new Pair(val[i] + val[i + 1], i));
            if (val[i] > val[i + 1]) inversions++;
        }

        int operations = 0;
        while (inversions > 0 && !pq.isEmpty()) {
            Pair top = pq.poll();
            int i = top.index;

            if (removed[i] || next[i] == -1 || val[i] + val[next[i]] != top.sum) {
                continue;
            }

            int j = next[i];
            int p = prev[i];
            int nn = next[j];

            if (p != -1 && val[p] > val[i]) inversions--;
            if (val[i] > val[j]) inversions--;
            if (nn != -1 && val[j] > val[nn]) inversions--;

            val[i] = top.sum;
            removed[j] = true;
            next[i] = nn;
            if (nn != -1) prev[nn] = i;

            if (p != -1 && val[p] > val[i]) inversions++;
            if (nn != -1 && val[i] > val[nn]) inversions++;

            if (p != -1) pq.add(new Pair(val[p] + val[i], p));
            if (nn != -1) pq.add(new Pair(val[i] + val[nn], i));

            operations++;
        }

        return operations;
    }
}

File: 3772-minimum-pair-removal-to-sort-array-ii/minimum-pair-removal-to-sort-array-ii.cpp
class Solution {
public:
    int minimumPairRemoval(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return 0;

        vector<long long> val(n);
        for (int i = 0; i < n; i++) val[i] = nums[i];

        vector<int> left(n), right(n);
        for (int i = 0; i < n; i++) {
            left[i] = i - 1;
            right[i] = i + 1;
        }
        right[n - 1] = -1;

        using Pair = pair<long long, int>;
        priority_queue<Pair, vector<Pair>, greater<Pair>> pq;
        
        int inversions = 0;
        for (int i = 0; i < n - 1; i++) {
            pq.push({val[i] + val[i + 1], i});
            if (val[i] > val[i + 1]) inversions++;
        }

        int operations = 0;
        vector<bool> removed(n, false);

        while (inversions > 0) {
            auto [current_sum, i] = pq.top();
            pq.pop();

            if (removed[i] || right[i] == -1 || val[i] + val[right[i]] != current_sum) {
                continue;
            }

            int j = right[i];
            int prev = left[i];
            int next = right[j];

            if (prev != -1 && val[prev] > val[i]) inversions--;
            if (val[i] > val[j]) inversions--;
            if (next != -1 && val[j] > val[next]) inversions--;

            val[i] = current_sum;
            removed[j] = true;
            right[i] = next;
            if (next != -1) left[next] = i;

            if (prev != -1 && val[prev] > val[i]) inversions++;
            if (next != -1 && val[i] > val[next]) inversions++;

            if (prev != -1) pq.push({val[prev] + val[i], prev});
            if (next != -1) pq.push({val[i] + val[next], i});

            operations++;
        }

        return operations;
    }
};